src/com/carolinarollergirls/scoreboard/defaults/DefaultStatsModel.java

```diff
@@ -0,0 +1,456 @@
+import com.carolinarollergirls.scoreboard.Position;
+import com.carolinarollergirls.scoreboard.ScoreBoard;
+import com.carolinarollergirls.scoreboard.Skater;
+import com.carolinarollergirls.scoreboard.Stats;
+import com.carolinarollergirls.scoreboard.Team;
+import com.carolinarollergirls.scoreboard.event.ConditionalScoreBoardListener;
+import com.carolinarollergirls.scoreboard.event.ScoreBoardEvent;
+import com.carolinarollergirls.scoreboard.event.ScoreBoardListener;
+import com.carolinarollergirls.scoreboard.model.StatsModel;
+
+public class DefaultStatsModel extends DefaultScoreBoardEventProvider implements StatsModel
+{
+  public DefaultStatsModel(ScoreBoard sb) {
+    scoreBoard = sb;
+
+    scoreBoard.addScoreBoardListener(new ConditionalScoreBoardListener(Clock.class, Clock.ID_PERIOD, Clock.EVENT_NUMBER, periodNumberListener));
+    scoreBoard.addScoreBoardListener(new ConditionalScoreBoardListener(Clock.class, Clock.ID_JAM, Clock.EVENT_NUMBER, jamNumberListener));
+    scoreBoard.addScoreBoardListener(new ConditionalScoreBoardListener(Clock.class, Clock.ID_JAM, Clock.EVENT_RUNNING, true, jamStartListener));
+    scoreBoard.addScoreBoardListener(new ConditionalScoreBoardListener(Clock.class, Clock.ID_JAM, Clock.EVENT_RUNNING, false, jamStopListener));
+
+    scoreBoard.addScoreBoardListener(new ConditionalScoreBoardListener(Team.class, Team.EVENT_SCORE, teamEventListener));
+    scoreBoard.addScoreBoardListener(new ConditionalScoreBoardListener(Team.class, Team.EVENT_LAST_SCORE, teamEventListener));
+    scoreBoard.addScoreBoardListener(new ConditionalScoreBoardListener(Team.class, Team.EVENT_LEAD_JAMMER, teamEventListener));
+    scoreBoard.addScoreBoardListener(new ConditionalScoreBoardListener(Team.class, Team.EVENT_STAR_PASS, teamEventListener));
+    scoreBoard.addScoreBoardListener(new ConditionalScoreBoardListener(Team.class, Team.EVENT_TIMEOUTS, teamEventListener));
+    scoreBoard.addScoreBoardListener(new ConditionalScoreBoardListener(Team.class, Team.EVENT_OFFICIAL_REVIEWS, teamEventListener));
+
+    scoreBoard.addScoreBoardListener(new ConditionalScoreBoardListener(Skater.class, Skater.EVENT_POSITION, skaterEventListener));
+    scoreBoard.addScoreBoardListener(new ConditionalScoreBoardListener(Skater.class, Skater.EVENT_PENALTY_BOX, skaterEventListener));
+
+    reset();
+  }
+
+  public ScoreBoard getScoreBoard() { return scoreBoard; }
+  public String getProviderName() { return "Stats"; }
+  public Class<Stats> getProviderClass() { return Stats.class; }
+  public String getProviderId() { return "Stats"; }
+
+  public void reset() {
+    truncateAfterNPeriods(0);
+  }
+
+  protected ScoreBoardListener periodNumberListener = new ScoreBoardListener() {
+    public void scoreBoardChange(ScoreBoardEvent event) {
+      // If the period number has dropped, we need to delete periods.
+      Clock pc = scoreBoard.getClock(Clock.ID_PERIOD);
+      truncateAfterNPeriods(pc.getNumber());
```

Comment:
This means that if an SBO accidentally decreases the period number, all stats for the current period are deleted, doesn't it? (And the same for jams.) We should look for a different way of handling such input. (Given that the previous code had the same problem and a solution is not immediately obvious, I'd consider this acceptable for now.)

Reply:
Yes, it's what the existing code does. At it stands you'd be able to reconstruct it from the previous json save.